docs/proposal/1-authentication-support/README.md

```diff
@@ -0,0 +1,175 @@
+# kstone-api Supports Authentication
+
+## Motivation
+
+Kstone has become a popular solution for etcd management. Due to lack of authentication ability, it has safety vulnerability and is not recommended using in production environment. To improve safety features of kstone, we'd like to support authentication in kstone.
+
+## Goals
+
+1. Use username/password authentication to implement login feature(create an admin user by default). 
+2. Support extendable authentication protocol (e.g. ldap, oidc).
+ 
+## Non-Goals
+
+Implement a complex RBAC authentication system.
+
+## Use Cases
+
+Visiting [kstone-dashboard](https://github.com/tkestack/kstone-dashboard) home page, it redirects to the login page.
+
+### Basic Auth
+
+Users use their username and password to login.
+
+### External Auth
+
+Users configure their own authentication system (e.g. ldap, oidc), and use their custom authentication system to verify identity.
+
+## Proposal
+
+In order to implement login feature, we develop an authentication module in kstone-api. It supports not only basic auth(username/password authentication) by default but also other authentication protocols.
+
+## Design Details
+
+### Authentication Process
+
+#### 1. Login Process
+
+![Login Process](./login.png)
+
+- First time login, POST request will be sent to kstone-api, URL path `/apis/login`.
+
+Request Example:
+
+```json
+{
+  "user":"admin",
+  "password":"admin"
+}
+```
+- **Authenticator**: Compare user and password with `kstone-api-admin` secret configuration.
+
+- **JWTTokenAuthenticator**: Generate a JWT token for users and put it in `set-cookie` response header.
+
+#### 2. Logout Process
+
+![Logout Process](./logout.png)
+
+- Clean cookies when logout.
+
+#### 3. JWT Token Authentication Process
+
+![Logout Process](./token-auth.png)
+
+- **JWTTokenAuthenticator**: Verify JWT Token cookie.
+
+> If user request with JWT token cookie, **JWTTokenAuthenticator** middleware will only verify the token until it is expired.
+
+### Authentication Interface
+
+```go
+
+// Token checks/generates a string value against a backing authentication store and
+// returns a Response or an error if the token could not be checked/created.
+type Token interface {
+    AuthenticateToken(ctx context.Context, token string) (*Response, bool, error)
+    GenerateToken(ctx context.Context, username string, password string) (string, error)
```

Comment:
We'd better define the TokenGenerator interface. The later extended authenticators do not necessarily need to implement this method. Most of them can reuse jwt tokens.

type TokenGenerator interface {
	// GenerateToken generates a token
	GenerateToken(...) (string, error)
}